R/correlation.R

```diff
@@ -0,0 +1,181 @@
+      }
+      if (!(int_2 %in% private$interventions)) {
+        stop(paste0('invalid intervention name: ', int_2))
+      }
+      if (rho < -1 || rho > 1) {
+        stop(paste0(
+          'rho between ',
+          int_1,
+          ' and ',
+          int_2,
+          'must be between -1 and 1'
+        ))
+      }
+      private$rho_matrix[[int_1, int_2]] <- rho
+      private$rho_matrix[[int_2, int_1]] <- rho
+      private$.sigma <- NULL
+      private$.mvnorm <- NULL
+    },
+
+    #' @description Standard deviation of each intervention between rounds
+    sigma = function() {
+      if (is.null(private$.sigma)) {
+        rho <- private$rho()
+        private$.sigma <- sqrt(rho / (1 - rho))
+        names(private$.sigma) <- private$interventions
+      }
+      private$.sigma
+    },
+
+    #' @description multivariate norm draws for these parameters
+    mvnorm = function() {
+      if (is.null(private$.mvnorm)) {
+        sigma <- self$sigma()
+        V <- outer(sigma, sigma) * private$rho_matrix
+        diag(V) <- sigma ^ 2
+        private$.mvnorm <- MASS::mvrnorm(
+          private$population,
+          rep(0, length(private$interventions)),
+          V
+        )
+        dimnames(private$.mvnorm)[[2]] <- private$interventions
+      }
+      private$.mvnorm
+    }
+  )
+)
+
+#' @title Get default correlation parameters
+#' @description returns a `CorrelationParameters` object for you edit. By
+#' default, all correlations are set to 0
+#'
+#' @param parameters model parameters
+#' @export
+#' @examples
+#' 
+#' # get the default model parameters
+#' parameters <- get_parameters()
+#' 
+#' # Set some rtss strategy
+#' parameters <- set_rtss(
+#'   parameters,
+#'   start = 100,
+#'   end = 1000,
+#'   frequency = 100,
+#'   min_ages = 100,
+#'   max_ages = 1000,
+#'   boosters = NULL,
+#'   coverage = .9,
+#'   booster_coverage = NULL
+#' )
+#' 
+#' # Set some smc strategy
+#' parameters <- set_drugs(parameters, list(SP_AQ_params))
+#' parameters <- set_smc(
+#'   parameters,
+#'   drug = 1,
+#'   start = 100,
+#'   end = 1000,
+#'   frequency = 100,
+#'   min_age = 100,
+#'   max_age = 1000,
+#'   coverage = .9
+#' )
+#' 
+#' # Correlate the rtss and smc targets
+#' c_param <- get_correlation_parameters(parameters)
+#' c_param$inter_intervention_rho('rtss', 'smc', 1)
+#' 
+#' # Correlate the rounds of smc
+#' c_param$inter_round_rho('smc', 1)
+#' 
+#' # You can now pass the correlation parameters to the run_simulation function
+get_correlation_parameters <- function(parameters) {
+  CorrelationParameters$new(parameters)
+}
+
+#' @title Sample a population to intervene in given the correlation parameters
+#' @param target a vector of individual indices to sample from
+#' @param intervention name of the intervention
+#' @param p the probability of being selected
+#' @param c_param correlation parameters
+#' @importFrom stats qnorm
+sample_intervention <- function(target, intervention, p, c_param) {
```

Comment:
As far as I understand the maths (not in much depth!), this seems to be replicating the Griffin 2010 approach well. My guess is that Patrick my have the best grasp on this if you wanted to ask someone to double-check.